tt_metal/impl/program/program.cpp

```diff
@@ -1453,6 +1462,16 @@ const std::vector<SubDeviceId> &detail::Program_::determine_sub_device_ids(const
     return sub_device_ids->second;
 }
 
+void detail::Program_::allocate_kernel_bin_buf_on_device(Device *device) {
+    // Allocate the DRAM kernel binary buffer for this program on the specified device, if not previously allocated.
+    // We allocate program binaries top down to minimize fragmentation with other buffers in DRAM, which are typically allocated bottom up
+    std::size_t binary_data_size_bytes = this->program_transfer_info.binary_data.size() * sizeof(uint32_t);
+    if (this->kernels_buffer.find(device->id()) == this->kernels_buffer.end() and binary_data_size_bytes) {
+        std::shared_ptr<Buffer> kernel_bin_buf = Buffer::create(device, binary_data_size_bytes, HostMemDeviceCommand::PROGRAM_PAGE_SIZE, BufferType::DRAM, TensorMemoryLayout::INTERLEAVED, std::nullopt, false);
+        this->kernels_buffer.insert({device->id(), kernel_bin_buf});
```

Comment:
Here as well. Btw checking for `program_transfer_info.empty()` in the `if` sounds a bit more natural, wdyt?

Reply:
Modified to use the `[]` accessor here. I'm not sure what this means:
`Btw checking for program_transfer_info.empty() in the if sounds a bit more natural, wdyt?`

Reply:
As in, instead of calculating `binary_data_size_bytes` upfront, express the `if` condition on whether `program_transfer_info.binary_data` is empty or not. Nit - feel free to ignore.